libbeat/cmd/instance/locks/lock.go

```diff
@@ -18,104 +18,64 @@
 package locks
 
 import (
-	"encoding/json"
-	"errors"
 	"fmt"
 	"os"
-	"runtime"
 	"time"
 
 	"github.com/gofrs/flock"
 
 	"github.com/elastic/beats/v7/libbeat/beat"
 	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/elastic/elastic-agent-libs/paths"
-	metricproc "github.com/elastic/elastic-agent-system-metrics/metric/system/process"
 )
 
+// Locker is a retrying file locker
 type Locker struct {
-	fileLock  *flock.Flock
-	logger    *logp.Logger
-	beatName  string
-	filePath  string
-	beatStart time.Time
-}
-
-type pidfile struct {
-	PID       int       `json:"pid"`
-	WriteTime time.Time `json:"write_time"`
+	fileLock   *flock.Flock
+	retryCount int
+	retrySleep time.Duration
+	logger     *logp.Logger
 }
 
 var (
 	// ErrAlreadyLocked is returned when a lock on the data path is attempted but
 	// unsuccessful because another Beat instance already has the lock on the same
 	// data path.
 	ErrAlreadyLocked = fmt.Errorf("data path already locked by another beat. Please make sure that multiple beats are not sharing the same data path (path.data)")
-
-	// ErrLockfileEmpty is returned by readExistingPidfile() when an existing pidfile is found, but the file is empty.
-	ErrLockfileEmpty = fmt.Errorf("lockfile is empty")
 )
 
-// a little wrapper for the gitpid function to make testing easier.
-var pidFetch = os.Getpid
-
 // New returns a new pid-aware file locker
-// all logic, including checking for existing locks, is performed lazily
 func New(beatInfo beat.Info) *Locker {
+	return NewWithRetry(beatInfo, 4, time.Millisecond*400)
+}
+
+// NewWithRetry returns a new pid-aware file locker with the given settings
+func NewWithRetry(beatInfo beat.Info, retryCount int, retrySleep time.Duration) *Locker {
 	lockfilePath := paths.Resolve(paths.Data, beatInfo.Beat+".lock")
 	return &Locker{
-		fileLock:  flock.New(lockfilePath),
-		logger:    logp.L(),
-		beatName:  beatInfo.Beat,
-		filePath:  lockfilePath,
-		beatStart: beatInfo.StartTime,
+		fileLock:   flock.New(lockfilePath),
+		retryCount: retryCount,
+		retrySleep: retrySleep,
+		logger:     logp.L(),
 	}
 }
 
 // Lock attempts to acquire a lock on the data path for the currently-running
 // Beat instance. If another Beats instance already has a lock on the same data path
 // an ErrAlreadyLocked error is returned.
 func (lock *Locker) Lock() error {
-	new := pidfile{PID: pidFetch(), WriteTime: time.Now()}
-	encoded, err := json.Marshal(&new)
-	if err != nil {
-		return fmt.Errorf("error encoding json for pidfile: %w", err)
-	}
-
-	// The combination of O_CREATE and O_EXCL will ensure we return an error if we don't
-	// manage to create the file
-	fh, openErr := os.OpenFile(lock.filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o600)
-	// Don't trust different OSes to report the errors we expect, just try to recover regardless
-	if openErr != nil {
-		err = lock.handleFailedCreate()
+	for i := 0; i < lock.retryCount; i++ {
+		gotLock, err := lock.fileLock.TryLock()
```

Comment:
I don't think TryLock uses the `os.O_EXCL` flag.  That means the file could exist already, and I think that would lead to a race condition in the Unlock function with Unlock & Remove.

Reply:
Could you elaborate? I assume you mean another beat swooping in while one beat is trying to lock or remove the file?

Reply:
In the unlock code, it first unlocks, then does a remove.  In between those lines of code another beat could create a new lock, but the file would be removed.  This results in the new beat having an error if it goes to unlock because the lock file doesn't exist.

```
panic: unable to unlock data path file testing.lock: remove testing.lock: no such file or directory
```

Reply:
hmm, that's an interesting edge case. Gonna see if I can think of a non-awkward way to protect against that.

Reply:
Alright, made a change to remove the file first before we remove the lock. Going to see how the Windows CI reacts to that, but I imagine we'll want some manual testing, since I don't understand the Windows lockfile logic too well.

Reply:
Can we put some or all of the detail from the PR description directly in the description of the Lock function? For example adding this would help the next developer to understand how this works.

> In a case where a beat has shutdown improperly and the lockfile remains, instead of looking up a PID, we rely on the OS to release the underlying lock for the dead process, which most OSes will generally do, after a set amount of time.

It may also be worth noting that putting the PID into the lock file failed. To some degree we have had several iterations on this code because the original code did not explain itself at all, so let's try to avoid creating that problem again.